### PR TITLE
TransitionRadiation plugin openPMD output

### PR DESCRIPTION
**Pull request by finnolec**
This PR adds openPMD output to the Transition Radiation plugin. Additionally, as a Quality of Life feature, it made the virtual foil position a command line parameter and it changed the plugin to a multi-plugin.

During the refactoring I noticed, that the current value storage implementation is quite messy with the internal 3D to 1D mapping, but this is a task for a future PR. Also this plugin would profit, as soon as the openPMD standard allows axes with different unit systems. The plugin with the openPMD output is probably less flexible in defining the observers now, this might need to be changed in the future as well.

@BeyondEspresso @psychocoderHPC

### CODE REVIEW

File: share/picongpu/examples/TransitionRadiation/etc/picongpu/16.cfg
Comment (PrometheusPi) [inline share/picongpu/examples/TransitionRadiation/etc/picongpu/16.cfg]: Is a foil at 0 useful?

File: share/picongpu/examples/TransitionRadiation/etc/picongpu/1.cfg
Comment (PrometheusPi) [inline share/picongpu/examples/TransitionRadiation/etc/picongpu/1.cfg]: is a foil at 0 useful?

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.kernel
Comment (PrometheusPi) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.kernel]: Just for my understanding, why do you subtract  `globalOffset[1]`?

File: share/picongpu/examples/TransitionRadiation/etc/picongpu/1.cfg
Comment (PrometheusPi) [reply share/picongpu/examples/TransitionRadiation/etc/picongpu/1.cfg]: Doesn't that mean `0==disabled`?

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (PrometheusPi) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: Just a reminder for me: this plugin does not require a restart routine.

**Review by PrometheusPi — DISMISSED** [review]
looks good to me. Just a question regarding the default foil position in the cfg files.

File: share/picongpu/examples/TransitionRadiation/etc/picongpu/1.cfg
Comment (finnolec) [reply share/picongpu/examples/TransitionRadiation/etc/picongpu/1.cfg]: No, a foil at 0 is physically useless, so I have this is the "the feature is disabled"-value. The point of this virtual foil propagation is, that one can naively "simulate" the vacuum-propagation of an accelerated electron bunch from an LWFA after the plasma stage, without having to run the simulation millimeters to centimeters that would mimic the experimental setup. The evolution of the bunch during the plasma stage is already handled much better by the PIC loop itself, if you want to know how the CTR spectra look before the current bunch position, you just call the plugin earlier. 
I was struggling a bit with what the clearest way is to have this optional parameter. Maybe I should add a new parameter called `--e_transRad.foilEnabled 0` to separate the activation and the parameter of this feature.

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (finnolec) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: Should I add this as a comment?

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.kernel
Comment (finnolec) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.kernel]: This took me some minutes to figure it out myself. I think, the idea is, that if we are in the simulation at `y=4mm`, but the foil is set to `foilPosition=10mm`, we only propagate the electrons 6mm instead of 10mm from their current position to the foil position. For this, it is using the simulation box position `globalOffest[1]` instead of the exact position of the single particles to not destroy the longitudinal/temporal coherence effects between the particles themselves.

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (psychocoderHPC) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: The naming should reflect that the position is for the Y-direction.

**Review by psychocoderHPC — DISMISSED** [review]

**Comment by franzpoeschel** [comment]
I can have a look at this after my holiday (Sep 4th)

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (franzpoeschel) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: Is there a specific reason why this would be a unique pointer?

I see two possible ways to improve the memory footprint here:

1. Use the Span API. This works only if use of the HostBuffer is not strictly necessary, because the Span API lets openPMD allocate a buffer. (`auto span =  transitionRadiation.storeChunk(offset, extent); span.currentBuffer()[index] = "<your data here>";`)
2. Use the unique-ptr based storeChunk overload, e.g. instead of passing a shared pointer of float to `storeChunk()`, pass a unique pointer. The IO backend can then potentially use the knowledge that it is the unique owner for optimizations (especially: the BP5 engine of ADIOS2 can avoid one data copy this way)

Do tell me if any of these options might work, then I can help you prepare such a workflow.

If not, then please consider using `storeChunkRaw()` from `include/picongpu/plugins/common/openPMDVersion.def` instead of creating a non-owning shared pointer.

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (franzpoeschel) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: I take it that this is the use case you mentioned a while ago where the mesh is not purely spatial? Otherwise, this specifies that one cell is a 1m cube.

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (franzpoeschel) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: See comment above, I recommend changing this either to (1) ideally a Span-based storeChunk(), (2) as a slightly weaker optimization a unique-ptr based storeChunk(), (3) or if those don't work a raw-pointer based storeChunk().

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (franzpoeschel) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]:
```suggestion
                    auto iteration = series.writeIterations()[currentStep];
                    auto mesh = iteration.meshes["transitionradiation"];
```

This API has better support for ADIOS2 steps.

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (finnolec) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: Yes, this grid has different units depending on the axis and I tried to use the radiation-plugin approach by having a separate grid for every axis, which determines the unit.

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (PrometheusPi) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: Not necessarily – as @psychocoderHPC code is our documentation, thus no code means no functionality – I am fine with that

File: docs/source/usage/plugins/transitionRadiation.rst
Comment (PrometheusPi) [inline docs/source/usage/plugins/transitionRadiation.rst]:
```suggestion
``--<species>_transRad.ext``               openPMD filename extension. Default: bp
```
or just write that it use the default used intrenally.

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (PrometheusPi) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: only suggestion: how about `po::bool_switch` instead?

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (PrometheusPi) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: why is this now an `init()` and no longer `pluginLoad()`?

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (PrometheusPi) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: shouldn't that be `1/UNIT_TIME`?

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (PrometheusPi) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: is `phi.setUnitDimension(...)` be default all zeros?

**Review by PrometheusPi — DISMISSED** [review]
looks good to me - just a few questions - and I will do a run time test before merging

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (franzpoeschel) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: Discussed today: We will leave this for now as I plan to implement https://github.com/openPMD/openPMD-standard/pull/193 soon which will help specifying this more precisely.

**Comment by PrometheusPi** [comment]
@finnolec what is the status of this pull request? 
may I test it or will there still be major changes and I should wait?

**Comment by finnolec** [comment]
I'm currently busy with different things, I will look into this PR in February again. I guess, you can wait until then with testing.

**Comment by psychocoderHPC** [comment]
@finnolec  please finish this PR or can we close it?

File: docs/source/usage/plugins/transitionRadiation.rst
Comment (finnolec) [reply docs/source/usage/plugins/transitionRadiation.rst]: done

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (finnolec) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: It was previously `pluginLoad()` because this is an inherited and implemented function in `ILightweightPlugin`. By switching to `plugins::multi::IInstance`, this function does not exist anymore. I prefer `init` slightly over `pluginLoad` and it's probably clearer, that this is now not function from the parent class

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (finnolec) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: What would be the advantage?

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (finnolec) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: The frequencies given by `transitionRadiation::frequencies::FreqFunctor` are not normalized, but in SI. So it should just be 1.0

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (finnolec) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: Offline discussion: It doesn't work for multi plugins

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (finnolec) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: I thought so, but I've clarified it now in the code :)

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (finnolec) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: done

**Comment by finnolec** [comment]
@PrometheusPi I implemented your suggestions, do you want to have another look?

**Comment by chillenzer** [comment]
Who would be capable of answering your last question? Maybe @PrometheusPi?

**Comment by steindev** [comment]
ping @PrometheusPi

**Comment by psychocoderHPC** [comment]
Will this PR ever be finished? If there is no update within the next two weeks I will close this PR it is open since 2023.

**Comment by PrometheusPi** [comment]
I will have a look today.

**Comment by PrometheusPi** [comment]
There were some issues with the CI and restart (and resubmit) was needed - I will have a look as soon as the CI gives a green light.

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.x.cpp
Comment (PrometheusPi) [inline include/picongpu/plugins/transitionRadiation/TransitionRadiation.x.cpp:229]: For some reason, one of the CI compilers (*g++-10_serial_boost1.78.0_examples.TransitionRadiation*) does not recognize the arguments on the next line and assumes that there are 0 instead of 3 arguments)

```
/builds/hzdr/crp/picongpu/include/picongpu/plugins/transitionRadiation/TransitionRadiation.x.cpp:229:17: note: candidate: 'picongpu::plugins::transitionRadiation::TransitionRadiation<ParticlesType>::TransitionRadiation(std::shared_ptr<picongpu::plugins::multi::IHelp>&, size_t, picongpu::MappingDesc*) [with ParticlesType = picongpu::Particles<pmacc::meta::String<'e'>, boost::mp11::mp_list<picongpu::particlePusher<picongpu::particles::pusher::Boris, pmacc::pmacc_isAlias>, picongpu::shape<picongpu::particles::shapes::TSC, pmacc::pmacc_isAlias>, picongpu::interpolation<picongpu::FieldToParticleInterpolation<picongpu::particles::shapes::TSC, picongpu::AssignedTrilinearInterpolation>, pmacc::pmacc_isAlias>, picongpu::massRatio<picongpu::MassRatioElectrons, pmacc::pmacc_isAlias>, picongpu::chargeRatio<picongpu::ChargeRatioElectrons, pmacc::pmacc_isAlias> >, boost::mp11::mp_list<picongpu::position<picongpu::position_pic, pmacc::pmacc_isAlias>, picongpu::momentum, picongpu::weighting, picongpu::transitionRadiationMask> >; size_t = long unsigned int; picongpu::MappingDesc = pmacc::MappingDescription<3, pmacc::math::CT::Vector<std::integral_constant<int, 8>, std::integral_constant<int, 8>, std::integral_constant<int, 4> > >]'
  229 |                 TransitionRadiation(
      |                 ^~~~~~~~~~~~~~~~~~~
/builds/hzdr/crp/picongpu/include/picongpu/plugins/transitionRadiation/TransitionRadiation.x.cpp:229:17: note:   candidate expects 3 arguments, 0 provided
make[2]: *** [CMakeFiles/picongpu.dir/build.make:440: CMakeFiles/picongpu.dir/plugins/transitionRadiation/TransitionRadiation.x.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:207: CMakeFiles/picongpu.dir/all] Error 2
```

Retriggering the CI causes the same error.

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
There is either an issue with the CI or with the code.

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.x.cpp
Comment (psychocoderHPC) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.x.cpp:229]: `PIC_REGISTER_SPECIES_PLUGIN(picongpu::plugins::multi::Master<picongpu::BinEnergyParticles<boost::mpl::_1>>);`

is an example how to use the macro at the end of this file. Each plugin must be constructable by `plugin()`

Comment (finnolec) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.x.cpp:229]: That was the error!

File: include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp
Comment (PrometheusPi) [reply include/picongpu/plugins/transitionRadiation/TransitionRadiation.hpp]: Thanks for the clarification @finnolec.

**Review by PrometheusPi — DISMISSED** [review]

**Comment by PrometheusPi** [comment]
@psychocoderHPC Your review is still negative. Should I overroule? 
@finnolec Your branch still has a merge conflict with our current `dev` and can thus not be merged - please fix.

**Comment by PrometheusPi** [comment]
Thanks @finnolec for fixing the merge conflict.

**Review by PrometheusPi — APPROVED** [review]

**Pull request merged**
